core/server/master/src/main/java/alluxio/underfs/UfsStatusCache.java

```diff
@@ -0,0 +1,282 @@
+   * @return the previous set of children if the mapping existed, null otherwise
+   */
+  @Nullable
+  public Collection<UfsStatus> addChildren(AlluxioURI path, Collection<UfsStatus> children) {
+    ConcurrentHashSet<UfsStatus> set = new ConcurrentHashSet<>();
+    children.forEach(child -> {
+      AlluxioURI childPath = path.joinUnsafe(child.getName());
+      addStatus(childPath, child);
+      set.add(child);
+    });
+    return mChildren.put(path, set);
+  }
+
+  /**
+   * Remove a status from the cache.
+   *
+   *  Any children added to this status will remain in the cache.
+   *
+   * @param path the path corresponding to the {@link UfsStatus} to remove
+   * @return the removed UfsStatus
+   */
+  public UfsStatus remove(AlluxioURI path) {
+    UfsStatus removed = mStatuses.remove(path);
+    if (removed == null) {
+      return null;
+    }
+
+    mChildren.remove(path); // ok if there aren't any children
+    return removed;
+  }
+
+  /**
+   * Get the UfsStatus from a given AlluxioURI.
+   *
+   * @param path the path the retrieve
+   * @return The corresponding {@link UfsStatus} or {@code null} if there is none stored
+   */
+  public UfsStatus getStatus(AlluxioURI path) {
+    return mStatuses.get(path);
+  }
+
+  /**
+   * Fetches children of a given alluxio path, stores them in the cache, then returns them.
+   *
+   * Children can be returned in a few ways
+   * 1. Children already exist in the internal index. We simply return them
+   * 2. If children did not already exist in the index, then check if there was a scheduled
+   * prefetch job running for this path. If so, wait for the job to finish and return the result.
+   * 3. If no prefetch job, and children don't yet exist in the cache, then if the fallback
+   * parameter is true, fetch them from the UFS and store them in the cache. Otherwise, simply
+   * return null.
+   *
+   * @param path the Alluxio path to get the children of
+   * @param mountTable the Alluxio mount table
+   * @param useFallback whether or not to fall back to calling the UFS
+   * @return child UFS statuses of the alluxio path, or null if no prefetch job and fallback
+   *         specified as false
+   * @throws InvalidPathException if the alluxio path can't be resolved to a UFS mount
+   */
+  @Nullable
+  public Collection<UfsStatus> fetchChildrenIfAbsent(AlluxioURI path, MountTable mountTable,
+      boolean useFallback)
+      throws IOException, InvalidPathException {
+    Future<Collection<UfsStatus>> prefetchJob = mActivePrefetchJobs.get(path);
+    if (prefetchJob != null) {
+      try {
+        return prefetchJob.get();
+      } catch (InterruptedException | ExecutionException e) {
+        LogUtils.warnWithException(LOG, "Failed to get result for prefetch job on alluxio path {}",
+            path, e);
+        if (e instanceof InterruptedException) {
+          Thread.currentThread().interrupt();
+        }
+        throw new IOException(e);
```

Comment:
If the prefetch job is canceled, this will throw an error, right? Does something external retry this?

Reply:
There is currently no retry.

I think that we should only throw the `IOException` if it is an InterruptedException, because that's something signalling the thread to stop. Otherwise if it is an ExecutionException or CancelledException then we should just continue with the fallback operation?